client/blocks/inline-help/data/use-inline-help-search-query.js

```diff
@@ -0,0 +1,69 @@
+import { createHigherOrderComponent } from '@wordpress/compose';
+import React from 'react';
+import { useQuery } from 'react-query';
+import { useSelector } from 'react-redux';
+import wpcom from 'calypso/lib/wp';
+import getAdminHelpResults from 'calypso/state/inline-help/selectors/get-admin-help-results';
+import getContextualHelpResults from 'calypso/state/inline-help/selectors/get-contextual-help-results';
+import getSearchQuery from 'calypso/state/inline-help/selectors/get-search-query';
+import { SUPPORT_TYPE_API_HELP, SUPPORT_TYPE_CONTEXTUAL_HELP } from '../constants';
+
+export const useInlineHelpSearchQuery = ( search, queryOptions = {} ) => {
+	const inlineHelpQuery = useQuery(
+		[ 'inline-help', search ],
+		async () => {
+			const { wordpress_support_links: searchResults } = await wpcom.req.get( '/help/search', {
+				query: search,
+				include_post_id: 1,
+			} );
+			return searchResults;
+		},
+		{ enabled: !! search, ...queryOptions }
+	);
+
+	return inlineHelpQuery;
+};
+
+/**
+ * Map the collection, populating each result object
+ * with the given support type value.
+ *
+ * @param {Array}   collection   - collection to populate.
+ * @param {string}  support_type - Support type to add to each result item.
+ * @returns {Array}                Populated collection.
+ */
+function mapWithSupportTypeProp( collection, support_type ) {
+	return collection.map( ( item ) => ( { ...item, support_type } ) );
+}
+
+export const withInlineHelpSearchQuery = createHigherOrderComponent(
+	( Wrapped ) => ( props ) => {
+		const searchQuery = useSelector( getSearchQuery );
+		const { data, isLoading } = useInlineHelpSearchQuery( searchQuery );
+
+		const hasApiResults = Array.isArray( data ) && data.length > 0;
+
+		const contextualResults = useSelector( getContextualHelpResults );
+		const mappedContextualResults = mapWithSupportTypeProp(
+			contextualResults,
+			SUPPORT_TYPE_CONTEXTUAL_HELP
+		);
+		const helpAdminResults = useSelector( ( state ) =>
+			getAdminHelpResults( state, searchQuery, 25 )
+		);
+
+		const searchResults = hasApiResults
+			? [ ...mapWithSupportTypeProp( data, SUPPORT_TYPE_API_HELP ), ...helpAdminResults ]
+			: [ ...( isLoading ? [] : mappedContextualResults ), ...helpAdminResults ];
```

Comment:
The code that creates the search results array could be greatly simplified. The API results, contextual results and admin results could be three completely independent array, retrieved as separate props by the component that shows the search results.

That `HelpSearchResults` component takes the `searchResults` array and divides it to three sub-arrays, one for each section. So why compose the merged array in the first place? The `HelpSearchResults` component could get the API results from the Query hook, the other two result arrays from independent selectors, and combine them together into a React UI only in the React UI component itself.

One obstacle to this is the `selectedResultIndex` state variable which is an index into the combined array, but #55833 should remove that obstacle.

Reply:
first iteration implemented in https://github.com/Automattic/wp-calypso/pull/55753/commits/e011d29420b319f06275c30911d9830c6ebd87dd